main.go

```diff
@@ -1,202 +1,272 @@
-				requiredDeploymentKeys: []types.NamespacedName{{
-					Name:      "cert-manager-webhook",
-					Namespace: "cert-manager",
-				}},
-				manifestName: "virtink.yaml",
-			}, {
-				crdName:      "clusters.cluster.x-k8s.io",
-				manifestName: "cluster-api-components.yaml",
-			}, {
-				crdName:      "virtinkclusters.infrastructure.cluster.x-k8s.io",
-				manifestName: "capch.yaml",
-			}}
-
-			for _, component := range components {
-				output, err := exec.Command("kubectl", "get", "crds", component.crdName, "--ignore-not-found").CombinedOutput()
-				if err != nil {
-					return fmt.Errorf("get CRD: %s", err)
-				}
-
-				if len(output) == 0 {
-					for _, k := range component.requiredDeploymentKeys {
-						if err := exec.Command("kubectl", "wait", "-n", k.Namespace, "deployments", k.Name, "--for=condition=Available", "--timeout=-1s").Run(); err != nil {
-							return fmt.Errorf("wait for deployment: %s", err)
-						}
-					}
-
-					if err := applyManifest(component.manifestName); err != nil {
-						return fmt.Errorf("apply manifest: %s", err)
-					}
-				}
-			}
-
-			capiWebhookDeploymentKeys := []types.NamespacedName{{
-				Name:      "capi-controller-manager",
-				Namespace: "capi-system",
-			}, {
-				Name:      "capi-kubeadm-bootstrap-controller-manager",
-				Namespace: "capi-kubeadm-bootstrap-system",
-			}, {
-				Name:      "capi-kubeadm-control-plane-controller-manager",
-				Namespace: "capi-kubeadm-control-plane-system",
-			}, {
-				Name:      "capch-controller-manager",
-				Namespace: "capch-system",
-			}}
-			for _, k := range capiWebhookDeploymentKeys {
-				if err := exec.Command("kubectl", "wait", "-n", k.Namespace, "deployments", k.Name, "--for=condition=Available", "--timeout=-1s").Run(); err != nil {
-					return fmt.Errorf("wait for deployment: %s", err)
-				}
-			}
-
-			if err := applyTemplate("cluster.yaml", map[string]interface{}{"name": clusterName}); err != nil {
-				return fmt.Errorf("apply cluster template: %s", err)
-			}
-
-			if err := exec.Command("kubectl", "wait", "clusters.cluster.x-k8s.io", clusterName, "--for=condition=ControlPlaneInitialized", "--timeout=-1s").Run(); err != nil {
-				return fmt.Errorf("wait for cluster: %s", err)
-			}
-
-			// TODO: change to kubectl
-			clusterService, err := kubeClient.CoreV1().Services("default").Get(cmd.Context(), clusterName, metav1.GetOptions{})
-			if err != nil {
-				return fmt.Errorf("get cluster service: %s", err)
-			}
-
-			clusterKubeconfigSecret, err := kubeClient.CoreV1().Secrets("default").Get(cmd.Context(), fmt.Sprintf("%s-kubeconfig", clusterName), metav1.GetOptions{})
-			if err != nil {
-				return fmt.Errorf("get cluster kubeconfig: %s", err)
-			}
-
-			clusterKubeconfigFileName := fmt.Sprintf("%s.kubeconfig", clusterName)
-			if err := os.WriteFile(clusterKubeconfigFileName, clusterKubeconfigSecret.Data["value"], 0644); err != nil {
-				return fmt.Errorf("save cluster kubeconfig: %s", err)
-			}
+		Use:   "init",
```

Comment:
So now we need two commands to create a cluster, knest init and then knest create xxx? If so, I don't think it's a nice user experience comparing to just use knest create xxx and everything is done

Reply:
a host cluster needs to be initialized only once, but user can create maybe workload clusters. furthermore user who init cluster maybe not the same one who create a workload cluster. and later we may add `knest upgrade` subcommand to upgrade cluster-api, virtink, virtink provider components.

Reply:
> a host cluster needs to be initialized only once, but user can create many workload clusters. 

True, we should detect if it's already initialized, and take that burden away from users.

> furthermore user who init cluster maybe not the same one who create a workload cluster. and later we may add knest upgrade subcommand to upgrade cluster-api, virtink, virtink provider components, this should not mixed with create cluster too.

knest is expected to be like minikube. For minikube, it doesn't require user to carefully separate between 'create' and 'start', just minikube start. That's a very nice thing. This should apply to knest too, just one command to take care all stuffs: init the cluster if necessary, create VMs if necessary, just start a K8s cluster that is ready to be used.

Reply:
that's a bit different. knest create a workload cluster on a host cluster, and cluster-api,  virtink, virtink provider are deployed on the host cluster which is public. Imagine a scenario, deploy cluster-api, virtink, virtink-providers is the job of Operations team, but create a workload cluster is for business team, business team just want a k8s cluster and they only have permissions to create a workload cluster, and they may never want to care about those components

Reply:
knest is to deal with the most straight forward case. For more sophisticated use cases, there will always be capch + clusterctl option